pkg/disk/disk.go

```diff
@@ -32,3 +44,106 @@ type Info struct {
 	// Usage is calculated per tenant.
 	Usage uint64
 }
+
+// Performance holds informantion about read and write speed of a disk
+type Performance struct {
+	Path       string  `json:"path"`
+	Error      string  `json:"error,omitempty"`
+	WriteSpeed float64 `json:"writeSpeed"`
+	ReadSpeed  float64 `json:"readSpeed"`
+}
+
+// GetPerformance returns given disk's read and write performance
+func GetPerformance(path string) Performance {
+	perf := Performance{}
+	write, err := writePerformance(path)
+	if err != nil {
+		perf.Error = err.Error()
+		return perf
+	}
+	read, err := readPerformance(path)
+	if err != nil {
+		perf.Error = err.Error()
+		return perf
+	}
+	perf.WriteSpeed = write
+	perf.ReadSpeed = read
+	return perf
+}
+
+// Calculate the write performance - write 10 tmp (1 MiB) files and
+// find the average time taken (Bytes / Sec)
+func writePerformance(fsPath string) (write float64, err error) {
+	var writeCount int
+	var totalElapsed time.Duration
+	buf := make([]byte, fileSize)
+	// Create a file 10 times in series, then find the average write time
+	for writeCount = 1; writeCount <= 10; writeCount++ {
+		rand.Read(buf)
+		tempObj := mustGetUUID()
+		fsTempObjPath := path.Join([]string{fsPath, tempObj}...)
+
+		start := time.Now()
+		n, err := writeFile(fsTempObjPath, buf)
+		elapsed := time.Since(start)
+
+		if err != nil {
+			removeFile(fsTempObjPath)
+			return 0, err
+		}
+
+		if n != fileSize {
+			removeFile(fsTempObjPath)
+			return 0, errors.New("Could not write temporary data to disk")
+		}
+
+		totalElapsed += elapsed
+		removeFile(fsTempObjPath)
+	}
+	// Average time spent = total time elapsed / number of writes
+	averageTime := totalElapsed.Seconds() / float64(writeCount)
+	// Write perf = fileSize (in Bytes) / average time spent (in seconds)
+	write = fileSize / averageTime
+	return write, nil
+}
+
+// Calculate the read performance - read a tmp (1 MiB) file 10 times and
+// find the average time taken (Bytes / Sec)
+func readPerformance(fsPath string) (read float64, err error) {
+	var readCount int
+	var totalElapsed time.Duration
+	tempObj := mustGetUUID()
+	fsTempObjPath := path.Join([]string{fsPath, tempObj}...)
+	buf := make([]byte, fileSize)
+	rand.Read(buf)
+	// Create a file to read
+	n, err := writeFile(fsTempObjPath, buf)
```

Comment:
Read shouldn't recreate a new file IMO, read and write should happen in the same function. This avoids you writing the file again.

Reply:
makes sense, done